src/vos/vos_dtx.c

```diff
@@ -102,179 +171,246 @@ dtx_hkey_cmp(struct btr_instance *tins, struct btr_record *rec, void *hkey)
-	.to_rec_update	= dtx_rec_update,
+	.to_rec_alloc	= dtx_act_ent_alloc,
+	.to_rec_free	= dtx_act_ent_free,
+	.to_rec_fetch	= dtx_act_ent_fetch,
+	.to_rec_update	= dtx_act_ent_update,
 };
 
-#define DTX_BTREE_ORDER		20
+static int
+dtx_cmt_ent_alloc(struct btr_instance *tins, d_iov_t *key_iov,
+		  d_iov_t *val_iov, struct btr_record *rec)
+{
+	struct vos_container	*cont = tins->ti_priv;
+	struct vos_dtx_cmt_ent	*dce = val_iov->iov_buf;
+
+	rec->rec_off = umem_ptr2off(&tins->ti_umm, dce);
+	if (!cont->vc_reindex_cmt_dtx || dce->dce_reindex) {
+		d_list_add_tail(&dce->dce_committed_link,
+				&cont->vc_dtx_committed_list);
+		cont->vc_dtx_committed_count++;
+	} else {
+		d_list_add_tail(&dce->dce_committed_link,
+				&cont->vc_dtx_committed_tmp_list);
+		cont->vc_dtx_committed_tmp_count++;
+	}
+
+	return 0;
+}
 
-int
-vos_dtx_table_register(void)
+static int
+dtx_cmt_ent_free(struct btr_instance *tins, struct btr_record *rec,
+		 void *args)
 {
-	int	rc;
+	struct vos_container	*cont = tins->ti_priv;
+	struct vos_dtx_cmt_ent	*dce;
+	int			 rc = 0;
 
-	D_DEBUG(DB_DF, "Registering DTX table class: %d\n", VOS_BTR_DTX_TABLE);
+	dce = umem_off2ptr(&tins->ti_umm, rec->rec_off);
+	D_ASSERT(dce != NULL);
 
-	rc = dbtree_class_register(VOS_BTR_DTX_TABLE, 0, &dtx_btr_ops);
-	if (rc != 0)
-		D_ERROR("Failed to register DTX dbtree: rc = %d\n", rc);
+	rec->rec_off = UMOFF_NULL;
+	d_list_del(&dce->dce_committed_link);
+	if (!cont->vc_reindex_cmt_dtx || dce->dce_reindex)
+		cont->vc_dtx_committed_count--;
+	else
+		cont->vc_dtx_committed_tmp_count--;
 
-	return rc;
-}
+	/* The committed DTX entry may be in aggregation now, if related
+	 * active DTX entry is waitting for batched cleanup, then cleanup
+	 * it by force before destroy the committed DTX entry (mainly for
+	 * DTX aggregation).
+	 */
+	if (!d_list_empty(&dce->dce_bcb_link)) {
+		struct umem_instance		*umm = vos_cont2umm(cont);
+		struct dtx_batched_cleanup_blob	*bcb = dce->dce_bcb;
+		struct vos_dtx_scm_blob		*dsb;
+		int				 idx;
 
-int
-vos_dtx_table_create(struct vos_pool *pool, struct vos_dtx_table_df *dtab_df)
-{
-	daos_handle_t	hdl;
-	int		rc;
+		D_ASSERT(bcb != NULL);
 
-	if (pool == NULL || dtab_df == NULL) {
-		D_ERROR("Invalid handle\n");
-		return -DER_INVAL;
-	}
+		idx = dce->dce_index;
+		D_ASSERT(idx >= 0);
 
-	D_ASSERT(dtab_df->tt_active_btr.tr_class == 0);
-	D_ASSERT(dtab_df->tt_committed_btr.tr_class == 0);
+		dsb = umem_off2ptr(umm, bcb->bcb_dsb_off);
 
-	D_DEBUG(DB_DF, "create DTX dbtree in-place for pool "DF_UUID": %d\n",
-		DP_UUID(pool->vp_id), VOS_BTR_DTX_TABLE);
+		if (dsb->dsb_active_data[idx].dae_flags != DTX_EF_INVALID) {
+			struct vos_dtx_act_ent_df	*dae_df;
 
-	rc = dbtree_create_inplace(VOS_BTR_DTX_TABLE, 0,
-				   DTX_BTREE_ORDER, &pool->vp_uma,
-				   &dtab_df->tt_active_btr, &hdl);
-	if (rc != 0) {
-		D_ERROR("Failed to create DTX active dbtree for pool "
-			DF_UUID": rc = %d\n", DP_UUID(pool->vp_id), rc);
-		return rc;
+			rc = vos_tx_begin(umm);
```

Comment:
we can start TX once and put everything in the same TX?
```
if (!tx_started) {
       rc = vos_tx_begin(umm);
       tx_started = true;
}
```

Reply:
What's mean? The outside caller may has already started the PMDK transaction, but here, we do not know, so we start the TX (again), if nested, it just increases the outside TX's reference that is handled by the PMDK logic. All the modification will belong to the same TX anyway.